{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import tidytcells as tt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TRA_CDR3</th>\n",
       "      <th>TRB_CDR3</th>\n",
       "      <th>Species</th>\n",
       "      <th>Category</th>\n",
       "      <th>Pathology</th>\n",
       "      <th>Pathology.Mesh.ID</th>\n",
       "      <th>Additional.study.details</th>\n",
       "      <th>Antigen.identification.method</th>\n",
       "      <th>Single.cell</th>\n",
       "      <th>NGS</th>\n",
       "      <th>...</th>\n",
       "      <th>TRAV</th>\n",
       "      <th>TRAJ</th>\n",
       "      <th>TRBV</th>\n",
       "      <th>TRBD</th>\n",
       "      <th>TRBJ</th>\n",
       "      <th>Reconstructed.J.annotation</th>\n",
       "      <th>CDR3.beta.nt</th>\n",
       "      <th>Mouse.strain</th>\n",
       "      <th>PubMed.ID</th>\n",
       "      <th>Remarks</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAVTIGFGNVLHCGSGTQVIVLPHIQ</td>\n",
       "      <td>CASSLGNEQF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>D008545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ35</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>No</td>\n",
       "      <td>TGTGCCAGCAGTCTTGGCAATGAGCAGTTCTTCGGGCCAGGGACAC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7777568</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CASGGGADGLTFPYIQF</td>\n",
       "      <td>CASSLGVATGELF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>D008545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-2</td>\n",
       "      <td>No</td>\n",
       "      <td>TGTGCCAGCAGTTTAGGGGTAGCGACCGGGGAGCTGTTTTTTGGAG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7777568</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAASPPESGGYNKLIF</td>\n",
       "      <td>CASSQEEGGGSWGNTIYF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>D008545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-3</td>\n",
       "      <td>No</td>\n",
       "      <td>TGTGCCAGCAGCCAAGAAGAGGGAGGAGGGTCTTGGGGAAACACCA...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7777568</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAAYYGGSOGNLIF</td>\n",
       "      <td>CASSQEGLAGASQYF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>D008545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-4</td>\n",
       "      <td>No</td>\n",
       "      <td>TGTGCCAGCAGCCAAGAGGGACTAGCGGGAGCGTCGCAGTACTTCG...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7777568</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAVSRGGGADGLTF</td>\n",
       "      <td>CASSQETDIVFNOPQHF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Cancer</td>\n",
       "      <td>Melanoma</td>\n",
       "      <td>D008545</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.4</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-5</td>\n",
       "      <td>No</td>\n",
       "      <td>TGCGCCAGCAGCCAAGAAACCGACATCGTCTTCAATCAGCCCCAGC...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7777568</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13696</th>\n",
       "      <td>CAGSGAGSYQLTF</td>\n",
       "      <td>CASSLEGQASSYEQYF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Pathogens</td>\n",
       "      <td>Epstein Barr virus (EBV)</td>\n",
       "      <td>D020031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35401538</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13697</th>\n",
       "      <td>CAGLGAGSYQLTF</td>\n",
       "      <td>CASSLEGQGASYEQYF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Pathogens</td>\n",
       "      <td>Epstein Barr virus (EBV)</td>\n",
       "      <td>D020031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35401538</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13698</th>\n",
       "      <td>CATEGNSGYSTLTF</td>\n",
       "      <td>CASSSQGGNYGYTF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Pathogens</td>\n",
       "      <td>Epstein Barr virus (EBV)</td>\n",
       "      <td>D020031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35401538</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13699</th>\n",
       "      <td>CATEGDSGYSTLTF</td>\n",
       "      <td>CASSYQGGNYGYTF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Pathogens</td>\n",
       "      <td>Epstein Barr virus (EBV)</td>\n",
       "      <td>D020031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35401538</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13700</th>\n",
       "      <td>CASEGNSGYSTLTF</td>\n",
       "      <td>CASSFQGGNYGYTF</td>\n",
       "      <td>HomoSapiens</td>\n",
       "      <td>Pathogens</td>\n",
       "      <td>Epstein Barr virus (EBV)</td>\n",
       "      <td>D020031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>...</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>35401538</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13701 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                         TRA_CDR3            TRB_CDR3      Species   Category  \\\n",
       "0      CAVTIGFGNVLHCGSGTQVIVLPHIQ          CASSLGNEQF  HomoSapiens     Cancer   \n",
       "1               CASGGGADGLTFPYIQF       CASSLGVATGELF  HomoSapiens     Cancer   \n",
       "2                CAASPPESGGYNKLIF  CASSQEEGGGSWGNTIYF  HomoSapiens     Cancer   \n",
       "3                  CAAYYGGSOGNLIF     CASSQEGLAGASQYF  HomoSapiens     Cancer   \n",
       "4                  CAVSRGGGADGLTF   CASSQETDIVFNOPQHF  HomoSapiens     Cancer   \n",
       "...                           ...                 ...          ...        ...   \n",
       "13696               CAGSGAGSYQLTF    CASSLEGQASSYEQYF  HomoSapiens  Pathogens   \n",
       "13697               CAGLGAGSYQLTF    CASSLEGQGASYEQYF  HomoSapiens  Pathogens   \n",
       "13698              CATEGNSGYSTLTF      CASSSQGGNYGYTF  HomoSapiens  Pathogens   \n",
       "13699              CATEGDSGYSTLTF      CASSYQGGNYGYTF  HomoSapiens  Pathogens   \n",
       "13700              CASEGNSGYSTLTF      CASSFQGGNYGYTF  HomoSapiens  Pathogens   \n",
       "\n",
       "                      Pathology Pathology.Mesh.ID Additional.study.details  \\\n",
       "0                      Melanoma           D008545                      NaN   \n",
       "1                      Melanoma           D008545                      NaN   \n",
       "2                      Melanoma           D008545                      NaN   \n",
       "3                      Melanoma           D008545                      NaN   \n",
       "4                      Melanoma           D008545                      NaN   \n",
       "...                         ...               ...                      ...   \n",
       "13696  Epstein Barr virus (EBV)           D020031                      NaN   \n",
       "13697  Epstein Barr virus (EBV)           D020031                      NaN   \n",
       "13698  Epstein Barr virus (EBV)           D020031                      NaN   \n",
       "13699  Epstein Barr virus (EBV)           D020031                      NaN   \n",
       "13700  Epstein Barr virus (EBV)           D020031                      NaN   \n",
       "\n",
       "       Antigen.identification.method Single.cell  NGS  ...       TRAV  \\\n",
       "0                                2.4          No   No  ...      TRAV2   \n",
       "1                                2.4          No   No  ...      TRAV2   \n",
       "2                                2.4          No   No  ...     TRAV21   \n",
       "3                                2.4          No   No  ...     TRAV21   \n",
       "4                                2.4          No   No  ...      TRAV2   \n",
       "...                              ...         ...  ...  ...        ...   \n",
       "13696                            1.0         Yes  Yes  ...     TRAV25   \n",
       "13697                            1.0         Yes  Yes  ...     TRAV25   \n",
       "13698                            1.0         Yes  Yes  ...  TRAV17*01   \n",
       "13699                            1.0         Yes  Yes  ...  TRAV17*01   \n",
       "13700                            1.0         Yes  Yes  ...  TRAV17*01   \n",
       "\n",
       "            TRAJ TRBV TRBD     TRBJ Reconstructed.J.annotation  \\\n",
       "0         TRAJ35  NaN  NaN  TRBJ2-1                         No   \n",
       "1         TRAJ45  NaN  NaN  TRBJ2-2                         No   \n",
       "2          TRAJ4  NaN  NaN  TRBJ1-3                         No   \n",
       "3         TRAJ42  NaN  NaN  TRBJ2-4                         No   \n",
       "4         TRAJ45  NaN  NaN  TRBJ1-5                         No   \n",
       "...          ...  ...  ...      ...                        ...   \n",
       "13696     TRAJ28  NaN  NaN      NaN                        NaN   \n",
       "13697     TRAJ28  NaN  NaN      NaN                        NaN   \n",
       "13698  TRAJ11*01  NaN  NaN      NaN                        NaN   \n",
       "13699  TRAJ11*01  NaN  NaN      NaN                        NaN   \n",
       "13700  TRAJ11*01  NaN  NaN      NaN                        NaN   \n",
       "\n",
       "                                            CDR3.beta.nt Mouse.strain  \\\n",
       "0      TGTGCCAGCAGTCTTGGCAATGAGCAGTTCTTCGGGCCAGGGACAC...          NaN   \n",
       "1      TGTGCCAGCAGTTTAGGGGTAGCGACCGGGGAGCTGTTTTTTGGAG...          NaN   \n",
       "2      TGTGCCAGCAGCCAAGAAGAGGGAGGAGGGTCTTGGGGAAACACCA...          NaN   \n",
       "3      TGTGCCAGCAGCCAAGAGGGACTAGCGGGAGCGTCGCAGTACTTCG...          NaN   \n",
       "4      TGCGCCAGCAGCCAAGAAACCGACATCGTCTTCAATCAGCCCCAGC...          NaN   \n",
       "...                                                  ...          ...   \n",
       "13696                                                NaN          NaN   \n",
       "13697                                                NaN          NaN   \n",
       "13698                                                NaN          NaN   \n",
       "13699                                                NaN          NaN   \n",
       "13700                                                NaN          NaN   \n",
       "\n",
       "      PubMed.ID Remarks  \n",
       "0       7777568     NaN  \n",
       "1       7777568     NaN  \n",
       "2       7777568     NaN  \n",
       "3       7777568     NaN  \n",
       "4       7777568     NaN  \n",
       "...         ...     ...  \n",
       "13696  35401538     NaN  \n",
       "13697  35401538     NaN  \n",
       "13698  35401538     NaN  \n",
       "13699  35401538     NaN  \n",
       "13700  35401538     NaN  \n",
       "\n",
       "[13701 rows x 29 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from_path = \"../../data/McPAS-TCR/fitted_data/mcpastcr_fitted_data_both.tsv\"\n",
    "mcpastcr_fitted_both_df = pd.read_csv(from_path, sep=\"\\t\")\n",
    "mcpastcr_fitted_both_df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TRA_CDR3', 'TRB_CDR3', 'Species', 'Category', 'Pathology',\n",
       "       'Pathology.Mesh.ID', 'Additional.study.details',\n",
       "       'Antigen.identification.method', 'Single.cell', 'NGS',\n",
       "       'Antigen.protein', 'Protein.ID', 'Epitope', 'Epitope.ID', 'MHC',\n",
       "       'Tissue', 'T.Cell.Type', 'T.cell.characteristics', 'CDR3.alpha.nt',\n",
       "       'TRAV', 'TRAJ', 'TRBV', 'TRBD', 'TRBJ', 'Reconstructed.J.annotation',\n",
       "       'CDR3.beta.nt', 'Mouse.strain', 'PubMed.ID', 'Remarks'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcpastcr_fitted_both_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcpastcr_cleaned_both_df = mcpastcr_fitted_both_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',\n",
       "       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',\n",
       "       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',\n",
       "       'TRB_3_prime_seq'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_columns = [\n",
    "    \"TCR_name\", 'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',\n",
    "    'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq', 'TRB_3_prime_seq'\n",
    "]\n",
    "\n",
    "for column in new_columns:\n",
    "    mcpastcr_cleaned_both_df[column] = np.nan\n",
    "\n",
    "template_columns_order = [\n",
    "    'TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',\n",
    "    'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',\n",
    "    'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq', 'TRB_3_prime_seq'\n",
    "]\n",
    "\n",
    "mcpastcr_cleaned_both_df = mcpastcr_cleaned_both_df[template_columns_order]\n",
    "\n",
    "mcpastcr_cleaned_both_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TCR_name', 'TRAV', 'TRAJ', 'TRA_CDR3', 'TRBV', 'TRBJ', 'TRB_CDR3',\n",
       "       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',\n",
       "       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',\n",
       "       'TRB_3_prime_seq', 'Epitope'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcpastcr_cleaned_both_df = pd.concat([mcpastcr_cleaned_both_df, mcpastcr_fitted_both_df[\"Epitope\"]], axis=1)\n",
    "mcpastcr_cleaned_both_df.columns\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(mcpastcr_fitted_both_df[\"T.Cell.Type\"].to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['TRA_CDR3', 'TRB_CDR3', 'Species', 'Category', 'Pathology',\n",
       "       'Pathology.Mesh.ID', 'Additional.study.details',\n",
       "       'Antigen.identification.method', 'Single.cell', 'NGS',\n",
       "       'Antigen.protein', 'Protein.ID', 'Epitope', 'Epitope.ID', 'MHC',\n",
       "       'Tissue', 'T.Cell.Type', 'T.cell.characteristics', 'CDR3.alpha.nt',\n",
       "       'TRAV', 'TRAJ', 'TRBV', 'TRBD', 'TRBJ', 'Reconstructed.J.annotation',\n",
       "       'CDR3.beta.nt', 'Mouse.strain', 'PubMed.ID', 'Remarks', 'TCR_name',\n",
       "       'TRAC', 'TRBC', 'TRA_leader', 'TRB_leader', 'Linker', 'Link_order',\n",
       "       'TRA_5_prime_seq', 'TRA_3_prime_seq', 'TRB_5_prime_seq',\n",
       "       'TRB_3_prime_seq'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define the mapping for the replacement\n",
    "value_mapping = {\n",
    "    \"CD8\": \"MHCI\",\n",
    "    \"CD4\": \"MHCII\"\n",
    "}\n",
    "\n",
    "# Use .map() to replace the values, passing the 'na_action='ignore'' parameter to keep NaN values as NaN\n",
    "mcpastcr_fitted_both_df[\"T.Cell.Type\"] = mcpastcr_fitted_both_df[\"T.Cell.Type\"].map(value_mapping, na_action='ignore')\n",
    "\n",
    "# Set anything else to np.nan\n",
    "# This step is only necessary if there are other values besides NaN, CD8, or CD4. If not, this can be skipped.\n",
    "# It's a safeguard in case your data changes over time or if you're unsure about the presence of other values.\n",
    "mcpastcr_fitted_both_df.loc[~mcpastcr_fitted_both_df[\"T.Cell.Type\"].isin([None, \"MHCI\", \"MHCII\"]), \"T.Cell.Type\"] = np.nan\n",
    "mcpastcr_fitted_both_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(mcpastcr_fitted_both_df[\"MHC\"].to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCR_name</th>\n",
       "      <th>TRAV</th>\n",
       "      <th>TRAJ</th>\n",
       "      <th>TRA_CDR3</th>\n",
       "      <th>TRBV</th>\n",
       "      <th>TRBJ</th>\n",
       "      <th>TRB_CDR3</th>\n",
       "      <th>TRAC</th>\n",
       "      <th>TRBC</th>\n",
       "      <th>TRA_leader</th>\n",
       "      <th>...</th>\n",
       "      <th>Link_order</th>\n",
       "      <th>TRA_5_prime_seq</th>\n",
       "      <th>TRA_3_prime_seq</th>\n",
       "      <th>TRB_5_prime_seq</th>\n",
       "      <th>TRB_3_prime_seq</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>Score</th>\n",
       "      <th>MHC A</th>\n",
       "      <th>MHC B</th>\n",
       "      <th>MHC class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ35</td>\n",
       "      <td>CAVTIGFGNVLHCGSGTQVIVLPHIQ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>CASSLGNEQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>CASGGGADGLTFPYIQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-2</td>\n",
       "      <td>CASSLGVATGELF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ4</td>\n",
       "      <td>CAASPPESGGYNKLIF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-3</td>\n",
       "      <td>CASSQEEGGGSWGNTIYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>CAAYYGGSOGNLIF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-4</td>\n",
       "      <td>CASSQEGLAGASQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>CAVSRGGGADGLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-5</td>\n",
       "      <td>CASSQETDIVFNOPQHF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13696</th>\n",
       "      <td>13697</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGSGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQASSYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13697</th>\n",
       "      <td>13698</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGLGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQGASYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13698</th>\n",
       "      <td>13699</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSSQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13699</th>\n",
       "      <td>13700</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGDSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13700</th>\n",
       "      <td>13701</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CASEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13701 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TCR_name       TRAV       TRAJ                    TRA_CDR3 TRBV  \\\n",
       "0             1      TRAV2     TRAJ35  CAVTIGFGNVLHCGSGTQVIVLPHIQ  NaN   \n",
       "1             2      TRAV2     TRAJ45           CASGGGADGLTFPYIQF  NaN   \n",
       "2             3     TRAV21      TRAJ4            CAASPPESGGYNKLIF  NaN   \n",
       "3             4     TRAV21     TRAJ42              CAAYYGGSOGNLIF  NaN   \n",
       "4             5      TRAV2     TRAJ45              CAVSRGGGADGLTF  NaN   \n",
       "...         ...        ...        ...                         ...  ...   \n",
       "13696     13697     TRAV25     TRAJ28               CAGSGAGSYQLTF  NaN   \n",
       "13697     13698     TRAV25     TRAJ28               CAGLGAGSYQLTF  NaN   \n",
       "13698     13699  TRAV17*01  TRAJ11*01              CATEGNSGYSTLTF  NaN   \n",
       "13699     13700  TRAV17*01  TRAJ11*01              CATEGDSGYSTLTF  NaN   \n",
       "13700     13701  TRAV17*01  TRAJ11*01              CASEGNSGYSTLTF  NaN   \n",
       "\n",
       "          TRBJ            TRB_CDR3  TRAC  TRBC  TRA_leader  ...  Link_order  \\\n",
       "0      TRBJ2-1          CASSLGNEQF   NaN   NaN         NaN  ...         NaN   \n",
       "1      TRBJ2-2       CASSLGVATGELF   NaN   NaN         NaN  ...         NaN   \n",
       "2      TRBJ1-3  CASSQEEGGGSWGNTIYF   NaN   NaN         NaN  ...         NaN   \n",
       "3      TRBJ2-4     CASSQEGLAGASQYF   NaN   NaN         NaN  ...         NaN   \n",
       "4      TRBJ1-5   CASSQETDIVFNOPQHF   NaN   NaN         NaN  ...         NaN   \n",
       "...        ...                 ...   ...   ...         ...  ...         ...   \n",
       "13696      NaN    CASSLEGQASSYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13697      NaN    CASSLEGQGASYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13698      NaN      CASSSQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13699      NaN      CASSYQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13700      NaN      CASSFQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "\n",
       "       TRA_5_prime_seq  TRA_3_prime_seq  TRB_5_prime_seq  TRB_3_prime_seq  \\\n",
       "0                  NaN              NaN              NaN              NaN   \n",
       "1                  NaN              NaN              NaN              NaN   \n",
       "2                  NaN              NaN              NaN              NaN   \n",
       "3                  NaN              NaN              NaN              NaN   \n",
       "4                  NaN              NaN              NaN              NaN   \n",
       "...                ...              ...              ...              ...   \n",
       "13696              NaN              NaN              NaN              NaN   \n",
       "13697              NaN              NaN              NaN              NaN   \n",
       "13698              NaN              NaN              NaN              NaN   \n",
       "13699              NaN              NaN              NaN              NaN   \n",
       "13700              NaN              NaN              NaN              NaN   \n",
       "\n",
       "          Epitope  Score     MHC A     MHC B MHC class  \n",
       "0      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "1      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "2      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "3      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "4      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "...           ...    ...       ...       ...       ...  \n",
       "13696   CLGGLLTMV    NaN       NaN       NaN       NaN  \n",
       "13697   CLGGLLTMV    NaN       NaN       NaN       NaN  \n",
       "13698   FLYALALLL    NaN       NaN       NaN       NaN  \n",
       "13699   FLYALALLL    NaN       NaN       NaN       NaN  \n",
       "13700   FLYALALLL    NaN       NaN       NaN       NaN  \n",
       "\n",
       "[13701 rows x 22 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcpastcr_cleaned_both_df[\"Score\"] = np.nan\n",
    "mcpastcr_cleaned_both_df[\"TCR_name\"] = range(1, len(mcpastcr_cleaned_both_df) + 1) # use index as TCR_name\n",
    "mcpastcr_cleaned_both_df[\"MHC A\"] = mcpastcr_fitted_both_df[\"MHC\"]\n",
    "mcpastcr_cleaned_both_df[\"MHC B\"] = mcpastcr_fitted_both_df[\"MHC\"]\n",
    "mcpastcr_cleaned_both_df[\"MHC class\"] = mcpastcr_fitted_both_df[\"T.Cell.Type\"]\n",
    "\n",
    "mcpastcr_cleaned_both_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_region_format(entry):\n",
    "    # Check if the entry is NaN or not a string to avoid errors.\n",
    "    if pd.isna(entry) or not isinstance(entry, str):\n",
    "        return entry  # Return the entry unchanged if it's NaN or not a string.\n",
    "    \n",
    "    # Only replace ':' with '*' if ':' is present in the entry.\n",
    "    # This specifically targets the allele annotation separation.\n",
    "    if ':' in entry:\n",
    "        entry = entry.replace(':', '*')\n",
    "    \n",
    "    return entry\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "for column in ['TRAV', 'TRAJ', 'TRBV', 'TRBJ']:\n",
    "    mcpastcr_cleaned_both_df[column] = mcpastcr_cleaned_both_df[column].apply(convert_region_format)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**check this...**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "def starts_with_m(df, columns):\n",
    "    mask = pd.Series([False] * len(df))\n",
    "    for column in columns:\n",
    "        temp_mask = df[column].apply(lambda entry: False if pd.isna(entry) else str(entry).startswith(\"m\"))\n",
    "        mask = mask | temp_mask\n",
    "    return mask\n",
    "\n",
    "columns_to_check = [\"TRAV\", \"TRBV\", \"TRAJ\", \"TRBJ\"]\n",
    "mask = starts_with_m(mcpastcr_cleaned_both_df, columns_to_check)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "mcpastcr_cleaned_both_df = mcpastcr_cleaned_both_df[~mask]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of columns to update\n",
    "columns_to_update = [\"TRAV\", \"TRBV\", \"TRAJ\", \"TRBJ\"]\n",
    "\n",
    "# Loop through each column and replace the pattern\n",
    "# This pattern finds '/D' followed by non-asterisk characters (if any) until an asterisk '*' or end of the string, but does not capture the asterisk and following characters for replacement\n",
    "for column in columns_to_update:\n",
    "    mcpastcr_cleaned_both_df.loc[:, column] = mcpastcr_cleaned_both_df[column].str.replace(r\"\\/D[^*]*\", \"\", regex=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# List of columns to update\n",
    "columns_to_update = [\"TRBV\", \"TRBJ\", \"TRAV\", \"TRAJ\"]\n",
    "\n",
    "# Loop through each column and replace everything after a ',' or '/' with an empty string\n",
    "for column in columns_to_update:\n",
    "    mcpastcr_cleaned_both_df.loc[:, column] = mcpastcr_cleaned_both_df[column].str.replace(r\"[,\\/].*\", \"\", regex=True)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCR_name</th>\n",
       "      <th>TRAV</th>\n",
       "      <th>TRAJ</th>\n",
       "      <th>TRA_CDR3</th>\n",
       "      <th>TRBV</th>\n",
       "      <th>TRBJ</th>\n",
       "      <th>TRB_CDR3</th>\n",
       "      <th>TRAC</th>\n",
       "      <th>TRBC</th>\n",
       "      <th>TRA_leader</th>\n",
       "      <th>...</th>\n",
       "      <th>Link_order</th>\n",
       "      <th>TRA_5_prime_seq</th>\n",
       "      <th>TRA_3_prime_seq</th>\n",
       "      <th>TRB_5_prime_seq</th>\n",
       "      <th>TRB_3_prime_seq</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>Score</th>\n",
       "      <th>MHC A</th>\n",
       "      <th>MHC B</th>\n",
       "      <th>MHC class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ35</td>\n",
       "      <td>CAVTIGFGNVLHCGSGTQVIVLPHIQ</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>CASSLGNEQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>CASGGGADGLTFPYIQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-2</td>\n",
       "      <td>CASSLGVATGELF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ4</td>\n",
       "      <td>CAASPPESGGYNKLIF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-3</td>\n",
       "      <td>CASSQEEGGGSWGNTIYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>CAAYYGGSOGNLIF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-4</td>\n",
       "      <td>CASSQEGLAGASQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>CAVSRGGGADGLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-5</td>\n",
       "      <td>CASSQETDIVFNOPQHF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13696</th>\n",
       "      <td>13697</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGSGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQASSYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13697</th>\n",
       "      <td>13698</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGLGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQGASYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13698</th>\n",
       "      <td>13699</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSSQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13699</th>\n",
       "      <td>13700</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGDSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13700</th>\n",
       "      <td>13701</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CASEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13661 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TCR_name       TRAV       TRAJ                    TRA_CDR3 TRBV  \\\n",
       "0             1      TRAV2     TRAJ35  CAVTIGFGNVLHCGSGTQVIVLPHIQ  NaN   \n",
       "1             2      TRAV2     TRAJ45           CASGGGADGLTFPYIQF  NaN   \n",
       "2             3     TRAV21      TRAJ4            CAASPPESGGYNKLIF  NaN   \n",
       "3             4     TRAV21     TRAJ42              CAAYYGGSOGNLIF  NaN   \n",
       "4             5      TRAV2     TRAJ45              CAVSRGGGADGLTF  NaN   \n",
       "...         ...        ...        ...                         ...  ...   \n",
       "13696     13697     TRAV25     TRAJ28               CAGSGAGSYQLTF  NaN   \n",
       "13697     13698     TRAV25     TRAJ28               CAGLGAGSYQLTF  NaN   \n",
       "13698     13699  TRAV17*01  TRAJ11*01              CATEGNSGYSTLTF  NaN   \n",
       "13699     13700  TRAV17*01  TRAJ11*01              CATEGDSGYSTLTF  NaN   \n",
       "13700     13701  TRAV17*01  TRAJ11*01              CASEGNSGYSTLTF  NaN   \n",
       "\n",
       "          TRBJ            TRB_CDR3  TRAC  TRBC  TRA_leader  ...  Link_order  \\\n",
       "0      TRBJ2-1          CASSLGNEQF   NaN   NaN         NaN  ...         NaN   \n",
       "1      TRBJ2-2       CASSLGVATGELF   NaN   NaN         NaN  ...         NaN   \n",
       "2      TRBJ1-3  CASSQEEGGGSWGNTIYF   NaN   NaN         NaN  ...         NaN   \n",
       "3      TRBJ2-4     CASSQEGLAGASQYF   NaN   NaN         NaN  ...         NaN   \n",
       "4      TRBJ1-5   CASSQETDIVFNOPQHF   NaN   NaN         NaN  ...         NaN   \n",
       "...        ...                 ...   ...   ...         ...  ...         ...   \n",
       "13696      NaN    CASSLEGQASSYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13697      NaN    CASSLEGQGASYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13698      NaN      CASSSQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13699      NaN      CASSYQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13700      NaN      CASSFQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "\n",
       "       TRA_5_prime_seq  TRA_3_prime_seq  TRB_5_prime_seq  TRB_3_prime_seq  \\\n",
       "0                  NaN              NaN              NaN              NaN   \n",
       "1                  NaN              NaN              NaN              NaN   \n",
       "2                  NaN              NaN              NaN              NaN   \n",
       "3                  NaN              NaN              NaN              NaN   \n",
       "4                  NaN              NaN              NaN              NaN   \n",
       "...                ...              ...              ...              ...   \n",
       "13696              NaN              NaN              NaN              NaN   \n",
       "13697              NaN              NaN              NaN              NaN   \n",
       "13698              NaN              NaN              NaN              NaN   \n",
       "13699              NaN              NaN              NaN              NaN   \n",
       "13700              NaN              NaN              NaN              NaN   \n",
       "\n",
       "          Epitope  Score     MHC A     MHC B MHC class  \n",
       "0      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "1      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "2      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "3      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "4      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "...           ...    ...       ...       ...       ...  \n",
       "13696   CLGGLLTMV    NaN       NaN       NaN       NaN  \n",
       "13697   CLGGLLTMV    NaN       NaN       NaN       NaN  \n",
       "13698   FLYALALLL    NaN       NaN       NaN       NaN  \n",
       "13699   FLYALALLL    NaN       NaN       NaN       NaN  \n",
       "13700   FLYALALLL    NaN       NaN       NaN       NaN  \n",
       "\n",
       "[13661 rows x 22 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcpastcr_cleaned_both_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV1-4\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV2-3\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV2-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV12-4\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV251\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV21-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV29*05\" for species homosapiens: nonexistent allele for recognised gene. (best attempted fix: \"TRAV29/DV5*05\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV78\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV3 F\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV3F\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV22 F\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV22F\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV8-1 F\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV8-1F\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV25 F\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV25F\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV41 F\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRAV41F\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV121\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV53\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV56\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRA1-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"36*01\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TR36*01\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV10-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAV19-*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRA21-01\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRA21-1\").\n",
      "  warn(warning_message)\n",
      "/tmp/ipykernel_23144/382921698.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mcpastcr_cleaned_both_df[col] = mcpastcr_cleaned_both_df[col].apply(standardize_gene)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV21-03\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV21-3\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV13-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV13-06\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV13-6\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV1-05\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV1-5\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV1-04\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV1-4\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-03\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV2-3\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-05\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV2-5\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV15-02\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV15-2\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV9-02\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV9-2\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV26-1*03\" for species homosapiens: nonexistent allele for recognised gene. (best attempted fix: \"TRBV26*03\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV8-4*05\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV1-5*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV1-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV1-4*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-6*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-7*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV19*04\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV19*06\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV26-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV19*05\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV2-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBV9-02*01\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRBV9-2*01\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ1-3\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ16-5\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ1-8\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ3-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ9-14\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ37-2\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ53*02\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ53*05\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJF\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRAJ36*02\" for species homosapiens: nonexistent allele for recognised gene.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"CATSESSGQTYEQYF\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRCAT-E--GQTYEQYF\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ1-7\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ5-6\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ53*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJTRBJ2-7\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ54*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ3-1*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ42*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ39*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ45*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ31*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ43*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ33*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ37*02\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ8*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ52*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ9*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ36*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"Donor 13\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRDON/OR13\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"Negative\" for species homosapiens: unrecognised gene name. (best attempted fix: \"TRNEGATIVE\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ5-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ10-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ20-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ19-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ17-1\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"TRBJ38-2*01\" for species homosapiens: unrecognised gene name.\n",
      "  warn(warning_message)\n"
     ]
    }
   ],
   "source": [
    "def standardize_gene(x):\n",
    "    if pd.notna(x): \n",
    "        return tt.tr.standardize(gene=x, species=\"homosapiens\")\n",
    "    else:\n",
    "        return x \n",
    "\n",
    "columns_gene = [\"TRAV\", \"TRBV\", \"TRAJ\", \"TRBJ\"]\n",
    "\n",
    "for col in columns_gene:\n",
    "    mcpastcr_cleaned_both_df[col] = mcpastcr_cleaned_both_df[col].apply(standardize_gene)\n",
    "\n",
    "# vdjdb_paired_cleaned_df[[\"TRAV\", \"TRBV\", \"TRAJ\", \"TRBJ\"]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAAYYGGSOGNLIF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CALSK(T)GALNLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CLLG(A)G(L)(D)AKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAGLGGT#GGSNYKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASKAAGNKLTF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CSRRAAGNKLTF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAFKAAGNKLTF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASDYGGSQGNLIF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize LGTGNQFYF : not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize YQLTF : not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize PYSSASKIIF : not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize STSFGSNYGQNFVF : not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAVN?PPFGNEKLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize TRAJ37: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/tmp/ipykernel_23144/2696340180.py:10: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mcpastcr_cleaned_both_df[col] = mcpastcr_cleaned_both_df[col].apply(standardize_CDR3)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSQETDIVFNOPQHF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRLAGGMDEQFF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRRTSGGTDTQYF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSGLAGGMDEQFF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSPGTSGVGEQFF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASRPLHSVYEQYF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CATTPGASGISEQFF*: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASRRVMARRE*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSL**LSKSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRLTPI*RNTIYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSW*PNEGKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASR*KVAHHSDNEQFF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASTHASRPE*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize LCQQKV*S*HE*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSPVSV*DKSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSQAVSG*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSKAEEM*QSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSH*ITISIRTEAFF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSDMGPI*KSYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAISIPGCDTK*GQRPDF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASRI*LLQHARNTIYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*GLPIYQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSNMIT*RKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAISGDRCR*KIYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*PYPFVMQDTQYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSS*V*GKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*HYEAGKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*TI*VHNEKLFF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSVLEG*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASV*GLGVQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSGCSP*HKSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSA*AHR*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSFGAGC*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASTHASRPE*KTYVYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSR*HV*DKSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*QIHFWHHQPQHF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSHGE*VQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAISGNSCR*KIYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASVQ*SVRKNYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSLAVRC*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CPCTLA*GEQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASKGRHICPE*KTYGDTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSGKGGC*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSI*CPIQRKNIQYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSMV*PE*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASR*GVATQFTEAFF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSMLG*I*QSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*RLATPSDNEQFF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSIGI*QSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*GIPICHEDTQYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*AHPFWHHQPQHF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASST*RLTQQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSTRRKG*KNYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSGRVSHFVLD*RYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSSP*RFAQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*GSPYDKSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRRSE*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSWGHV*GQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSISCT*RKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*RIKCHKSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CGKICR*RKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*L*RKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRGPR*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASR*QSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSLG*VPQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*GGKCQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSV*LRHE*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRC*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRVG*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize WQ*S*PIGGRRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAISFLSRR*KTHGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASIMYFA*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAS*IPISRKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSARLKS*RTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASKPN*EYQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSHHKRC*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSK*GPICRKNIQYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSTMEVC*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSV*HISGLTGELFF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAS*LS*LKQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSRVN*EPQSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASR*YLLQHARNTIYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize YASS*AFFDGKIYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASST*VPIVMQDTQYF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASS*PRYEGKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSPLVEM*QSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSW*LN*QSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSGALPGKTYG*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASREVVATQSH*SFL: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSDQRIVNKVGPTS*LS: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSLLLG*QSRANVLTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CGNGYR*RKTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSSRSTG*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CAAVGTLTHE*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSPDRHR*KTYGYTF: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize Donor13: not a valid amino acid sequence.\n",
      "  warnings.warn(\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/aa/_standardize.py:80: UserWarning: Failed to standardize CASSKVAARVP-TLKLS: not a valid amino acid sequence.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "def standardize_CDR3(x):\n",
    "    if pd.notna(x): \n",
    "        return tt.junction.standardize(seq=x)\n",
    "    else:\n",
    "        return x \n",
    "    \n",
    "columns_CDR3 = [\"TRA_CDR3\", \"TRB_CDR3\"]\n",
    "\n",
    "for col in columns_CDR3: \n",
    "    mcpastcr_cleaned_both_df[col] = mcpastcr_cleaned_both_df[col].apply(standardize_CDR3)\n",
    "\n",
    "\n",
    "# vdjdb_paired_cleaned_df[[\"TRA_CDR3\", \"TRB_CDR3\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"DR3*02:02\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DR3*02:02\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-DQ2\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DQ*2\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"H-2Kb\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-H-2KB\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-DQ8\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DQ*8\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"DQ8-trans\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DQ8-TRANS\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-DR1\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DR*1\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-DR11\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DR*11\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-DR15\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DR*15\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-DR5\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DR*5\").\n",
      "  warn(warning_message)\n",
      "/home/ubuntu/anaconda3/envs/BA24/lib/python3.12/site-packages/tidytcells/_utils/warnings.py:12: UserWarning: Failed to standardize \"HLA-DQ2.5\" for species homosapiens: unrecognised gene name. (best attempted fix: \"HLA-DQ*2:5\").\n",
      "  warn(warning_message)\n",
      "/tmp/ipykernel_23144/829523673.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mcpastcr_cleaned_both_df[col] = mcpastcr_cleaned_both_df[col].apply(standardize_MHC)\n"
     ]
    }
   ],
   "source": [
    "def standardize_MHC(x): \n",
    "    if pd.notna(x): \n",
    "        return tt.mh.standardize(gene=x, species=\"homosapiens\")\n",
    "    \n",
    "columns_MHC = [\"MHC A\", \"MHC B\"]\n",
    "\n",
    "for col in columns_MHC: \n",
    "    mcpastcr_cleaned_both_df[col] = mcpastcr_cleaned_both_df[col].apply(standardize_MHC)\n",
    "\n",
    "# print(vdjdb_paired_cleaned_df[[\"MHC A\", \"MHC B\"]].to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "paired_condition = mcpastcr_cleaned_both_df[\"TRA_CDR3\"].notna() & mcpastcr_cleaned_both_df[\"TRB_CDR3\"].notna()\n",
    "\n",
    "# Condition for beta only information (TRB_CDR3 is not NaN)\n",
    "beta_only_condition = mcpastcr_cleaned_both_df[\"TRB_CDR3\"].notna()\n",
    "\n",
    "# Creating DataFrames based on conditions\n",
    "mcpastcr_cleaned_paired_df = mcpastcr_cleaned_both_df[paired_condition]\n",
    "mcpastcr_cleaned_beta_df = mcpastcr_cleaned_both_df[beta_only_condition]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCR_name</th>\n",
       "      <th>TRAV</th>\n",
       "      <th>TRAJ</th>\n",
       "      <th>TRA_CDR3</th>\n",
       "      <th>TRBV</th>\n",
       "      <th>TRBJ</th>\n",
       "      <th>TRB_CDR3</th>\n",
       "      <th>TRAC</th>\n",
       "      <th>TRBC</th>\n",
       "      <th>TRA_leader</th>\n",
       "      <th>...</th>\n",
       "      <th>Link_order</th>\n",
       "      <th>TRA_5_prime_seq</th>\n",
       "      <th>TRA_3_prime_seq</th>\n",
       "      <th>TRB_5_prime_seq</th>\n",
       "      <th>TRB_3_prime_seq</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>Score</th>\n",
       "      <th>MHC A</th>\n",
       "      <th>MHC B</th>\n",
       "      <th>MHC class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ35</td>\n",
       "      <td>CCAVTIGFGNVLHCGSGTQVIVLPHIQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>CASSLGNEQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>CASGGGADGLTFPYIQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-2</td>\n",
       "      <td>CASSLGVATGELF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ4</td>\n",
       "      <td>CAASPPESGGYNKLIF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-3</td>\n",
       "      <td>CASSQEEGGGSWGNTIYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>TRAV10</td>\n",
       "      <td>TRAJ47</td>\n",
       "      <td>CAGAREYGNKLVF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>CASSQGQLTDTQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ35</td>\n",
       "      <td>CCAVNIGFGNVLHCGSGTQVIVLPHIQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>CASSRTVGGPNEQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13696</th>\n",
       "      <td>13697</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGSGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQASSYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13697</th>\n",
       "      <td>13698</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGLGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQGASYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13698</th>\n",
       "      <td>13699</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSSQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13699</th>\n",
       "      <td>13700</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGDSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13700</th>\n",
       "      <td>13701</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CASEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3041 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TCR_name       TRAV       TRAJ                      TRA_CDR3 TRBV  \\\n",
       "0             1      TRAV2     TRAJ35  CCAVTIGFGNVLHCGSGTQVIVLPHIQF  NaN   \n",
       "1             2      TRAV2     TRAJ45             CASGGGADGLTFPYIQF  NaN   \n",
       "2             3     TRAV21      TRAJ4              CAASPPESGGYNKLIF  NaN   \n",
       "5             6     TRAV10     TRAJ47                 CAGAREYGNKLVF  NaN   \n",
       "6             7      TRAV2     TRAJ35  CCAVNIGFGNVLHCGSGTQVIVLPHIQF  NaN   \n",
       "...         ...        ...        ...                           ...  ...   \n",
       "13696     13697     TRAV25     TRAJ28                 CAGSGAGSYQLTF  NaN   \n",
       "13697     13698     TRAV25     TRAJ28                 CAGLGAGSYQLTF  NaN   \n",
       "13698     13699  TRAV17*01  TRAJ11*01                CATEGNSGYSTLTF  NaN   \n",
       "13699     13700  TRAV17*01  TRAJ11*01                CATEGDSGYSTLTF  NaN   \n",
       "13700     13701  TRAV17*01  TRAJ11*01                CASEGNSGYSTLTF  NaN   \n",
       "\n",
       "          TRBJ            TRB_CDR3  TRAC  TRBC  TRA_leader  ...  Link_order  \\\n",
       "0      TRBJ2-1          CASSLGNEQF   NaN   NaN         NaN  ...         NaN   \n",
       "1      TRBJ2-2       CASSLGVATGELF   NaN   NaN         NaN  ...         NaN   \n",
       "2      TRBJ1-3  CASSQEEGGGSWGNTIYF   NaN   NaN         NaN  ...         NaN   \n",
       "5      TRBJ2-3      CASSQGQLTDTQYF   NaN   NaN         NaN  ...         NaN   \n",
       "6      TRBJ2-1      CASSRTVGGPNEQF   NaN   NaN         NaN  ...         NaN   \n",
       "...        ...                 ...   ...   ...         ...  ...         ...   \n",
       "13696      NaN    CASSLEGQASSYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13697      NaN    CASSLEGQGASYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13698      NaN      CASSSQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13699      NaN      CASSYQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13700      NaN      CASSFQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "\n",
       "       TRA_5_prime_seq  TRA_3_prime_seq  TRB_5_prime_seq  TRB_3_prime_seq  \\\n",
       "0                  NaN              NaN              NaN              NaN   \n",
       "1                  NaN              NaN              NaN              NaN   \n",
       "2                  NaN              NaN              NaN              NaN   \n",
       "5                  NaN              NaN              NaN              NaN   \n",
       "6                  NaN              NaN              NaN              NaN   \n",
       "...                ...              ...              ...              ...   \n",
       "13696              NaN              NaN              NaN              NaN   \n",
       "13697              NaN              NaN              NaN              NaN   \n",
       "13698              NaN              NaN              NaN              NaN   \n",
       "13699              NaN              NaN              NaN              NaN   \n",
       "13700              NaN              NaN              NaN              NaN   \n",
       "\n",
       "          Epitope  Score     MHC A     MHC B MHC class  \n",
       "0      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "1      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "2      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "5      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "6      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "...           ...    ...       ...       ...       ...  \n",
       "13696   CLGGLLTMV    NaN      None      None       NaN  \n",
       "13697   CLGGLLTMV    NaN      None      None       NaN  \n",
       "13698   FLYALALLL    NaN      None      None       NaN  \n",
       "13699   FLYALALLL    NaN      None      None       NaN  \n",
       "13700   FLYALALLL    NaN      None      None       NaN  \n",
       "\n",
       "[3041 rows x 22 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcpastcr_cleaned_paired_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCR_name</th>\n",
       "      <th>TRAV</th>\n",
       "      <th>TRAJ</th>\n",
       "      <th>TRA_CDR3</th>\n",
       "      <th>TRBV</th>\n",
       "      <th>TRBJ</th>\n",
       "      <th>TRB_CDR3</th>\n",
       "      <th>TRAC</th>\n",
       "      <th>TRBC</th>\n",
       "      <th>TRA_leader</th>\n",
       "      <th>...</th>\n",
       "      <th>Link_order</th>\n",
       "      <th>TRA_5_prime_seq</th>\n",
       "      <th>TRA_3_prime_seq</th>\n",
       "      <th>TRB_5_prime_seq</th>\n",
       "      <th>TRB_3_prime_seq</th>\n",
       "      <th>Epitope</th>\n",
       "      <th>Score</th>\n",
       "      <th>MHC A</th>\n",
       "      <th>MHC B</th>\n",
       "      <th>MHC class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ35</td>\n",
       "      <td>CCAVTIGFGNVLHCGSGTQVIVLPHIQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-1</td>\n",
       "      <td>CASSLGNEQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>TRAV2</td>\n",
       "      <td>TRAJ45</td>\n",
       "      <td>CASGGGADGLTFPYIQF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-2</td>\n",
       "      <td>CASSLGVATGELF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ4</td>\n",
       "      <td>CAASPPESGGYNKLIF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ1-3</td>\n",
       "      <td>CASSQEEGGGSWGNTIYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>TRAV21</td>\n",
       "      <td>TRAJ42</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-4</td>\n",
       "      <td>CASSQEGLAGASQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>6</td>\n",
       "      <td>TRAV10</td>\n",
       "      <td>TRAJ47</td>\n",
       "      <td>CAGAREYGNKLVF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TRBJ2-3</td>\n",
       "      <td>CASSQGQLTDTQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAAGIGILTV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>HLA-A*02</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13696</th>\n",
       "      <td>13697</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGSGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQASSYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13697</th>\n",
       "      <td>13698</td>\n",
       "      <td>TRAV25</td>\n",
       "      <td>TRAJ28</td>\n",
       "      <td>CAGLGAGSYQLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSLEGQGASYEQYF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CLGGLLTMV</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13698</th>\n",
       "      <td>13699</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSSQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13699</th>\n",
       "      <td>13700</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CATEGDSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSYQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13700</th>\n",
       "      <td>13701</td>\n",
       "      <td>TRAV17*01</td>\n",
       "      <td>TRAJ11*01</td>\n",
       "      <td>CASEGNSGYSTLTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CASSFQGGNYGYTF</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FLYALALLL</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12117 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       TCR_name       TRAV       TRAJ                      TRA_CDR3 TRBV  \\\n",
       "0             1      TRAV2     TRAJ35  CCAVTIGFGNVLHCGSGTQVIVLPHIQF  NaN   \n",
       "1             2      TRAV2     TRAJ45             CASGGGADGLTFPYIQF  NaN   \n",
       "2             3     TRAV21      TRAJ4              CAASPPESGGYNKLIF  NaN   \n",
       "3             4     TRAV21     TRAJ42                          None  NaN   \n",
       "5             6     TRAV10     TRAJ47                 CAGAREYGNKLVF  NaN   \n",
       "...         ...        ...        ...                           ...  ...   \n",
       "13696     13697     TRAV25     TRAJ28                 CAGSGAGSYQLTF  NaN   \n",
       "13697     13698     TRAV25     TRAJ28                 CAGLGAGSYQLTF  NaN   \n",
       "13698     13699  TRAV17*01  TRAJ11*01                CATEGNSGYSTLTF  NaN   \n",
       "13699     13700  TRAV17*01  TRAJ11*01                CATEGDSGYSTLTF  NaN   \n",
       "13700     13701  TRAV17*01  TRAJ11*01                CASEGNSGYSTLTF  NaN   \n",
       "\n",
       "          TRBJ            TRB_CDR3  TRAC  TRBC  TRA_leader  ...  Link_order  \\\n",
       "0      TRBJ2-1          CASSLGNEQF   NaN   NaN         NaN  ...         NaN   \n",
       "1      TRBJ2-2       CASSLGVATGELF   NaN   NaN         NaN  ...         NaN   \n",
       "2      TRBJ1-3  CASSQEEGGGSWGNTIYF   NaN   NaN         NaN  ...         NaN   \n",
       "3      TRBJ2-4     CASSQEGLAGASQYF   NaN   NaN         NaN  ...         NaN   \n",
       "5      TRBJ2-3      CASSQGQLTDTQYF   NaN   NaN         NaN  ...         NaN   \n",
       "...        ...                 ...   ...   ...         ...  ...         ...   \n",
       "13696      NaN    CASSLEGQASSYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13697      NaN    CASSLEGQGASYEQYF   NaN   NaN         NaN  ...         NaN   \n",
       "13698      NaN      CASSSQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13699      NaN      CASSYQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "13700      NaN      CASSFQGGNYGYTF   NaN   NaN         NaN  ...         NaN   \n",
       "\n",
       "       TRA_5_prime_seq  TRA_3_prime_seq  TRB_5_prime_seq  TRB_3_prime_seq  \\\n",
       "0                  NaN              NaN              NaN              NaN   \n",
       "1                  NaN              NaN              NaN              NaN   \n",
       "2                  NaN              NaN              NaN              NaN   \n",
       "3                  NaN              NaN              NaN              NaN   \n",
       "5                  NaN              NaN              NaN              NaN   \n",
       "...                ...              ...              ...              ...   \n",
       "13696              NaN              NaN              NaN              NaN   \n",
       "13697              NaN              NaN              NaN              NaN   \n",
       "13698              NaN              NaN              NaN              NaN   \n",
       "13699              NaN              NaN              NaN              NaN   \n",
       "13700              NaN              NaN              NaN              NaN   \n",
       "\n",
       "          Epitope  Score     MHC A     MHC B MHC class  \n",
       "0      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "1      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "2      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "3      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "5      EAAGIGILTV    NaN  HLA-A*02  HLA-A*02       NaN  \n",
       "...           ...    ...       ...       ...       ...  \n",
       "13696   CLGGLLTMV    NaN      None      None       NaN  \n",
       "13697   CLGGLLTMV    NaN      None      None       NaN  \n",
       "13698   FLYALALLL    NaN      None      None       NaN  \n",
       "13699   FLYALALLL    NaN      None      None       NaN  \n",
       "13700   FLYALALLL    NaN      None      None       NaN  \n",
       "\n",
       "[12117 rows x 22 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mcpastcr_cleaned_beta_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_path = \"../../data/McPAS-TCR/cleaned_data\"\n",
    "file_name = \"mcpastcr_cleaned_data_both.tsv\"\n",
    "mcpastcr_cleaned_both_df.to_csv(to_path+\"/\"+file_name, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_path = \"../../data/McPAS-TCR/cleaned_data\"\n",
    "file_name = \"mcpastcr_cleaned_data_paired.tsv\"\n",
    "mcpastcr_cleaned_paired_df.to_csv(to_path+\"/\"+file_name, sep=\"\\t\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_path = \"../../data/McPAS-TCR/cleaned_data\"\n",
    "file_name = \"mcpastcr_cleaned_data_beta.tsv\"\n",
    "mcpastcr_cleaned_beta_df.to_csv(to_path+\"/\"+file_name, sep=\"\\t\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BA24",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
