{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this file we look closer at the IEDB Data.\n",
    "The csv was created with the following filters:\n",
    "\n",
    "![alt text](image.png)\n",
    "\n",
    "After loading this additional filter was set:\n",
    "\n",
    "![alt text](image-1.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data has many columns, not all of them are interesting for our purposes.\n",
    "Now the data get's loaded and the columns are seperated in three lists. One list with interesting properties, one with properties which might be interesting and the last one with properties which is neither in list one or two."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_60458/2710885902.py:3: DtypeWarning: Columns (2,10,12,13,14,15,16,17,19,20,21,22,23,24,29,30,35,36,41,42,43,44,45,46,47,48,49,50,51,52,53,58,59,64,65) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  df = pd.read_csv(\"../../data/IEDB/receptor_table_export_1711285770.csv\")\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"../../data/IEDB/receptor_table_export_1711285770.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "now checkout the dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                Receptor - Group IRI  Receptor - IEDB Receptor ID  \\\n",
      "0    http://www.iedb.org/receptor/47                           57   \n",
      "1    http://www.iedb.org/receptor/47                           57   \n",
      "2  http://www.iedb.org/receptor/8493                           58   \n",
      "3  http://www.iedb.org/receptor/8493                           58   \n",
      "4  http://www.iedb.org/receptor/8493                           58   \n",
      "\n",
      "  Receptor - Reference Name Receptor - Type  \\\n",
      "0                    KK50.4       alphabeta   \n",
      "1                    KK50.4       alphabeta   \n",
      "2                      LC13       alphabeta   \n",
      "3                      LC13       alphabeta   \n",
      "4                      LC13       alphabeta   \n",
      "\n",
      "                    Reference - IEDB IRI                  Epitope - IEDB IRI  \\\n",
      "0  http://www.iedb.org/reference/1004539   http://www.iedb.org/epitope/69921   \n",
      "1  http://www.iedb.org/reference/1004539   http://www.iedb.org/epitope/69921   \n",
      "2  http://www.iedb.org/reference/1004580  http://www.iedb.org/epitope/144889   \n",
      "3  http://www.iedb.org/reference/1004580   http://www.iedb.org/epitope/16878   \n",
      "4  http://www.iedb.org/reference/1017865  http://www.iedb.org/epitope/156273   \n",
      "\n",
      "  Epitope - Name                          Epitope - Source Molecule  \\\n",
      "0      VMAPRTLIL  HLA class I histocompatibility antigen, Cw-3 a...   \n",
      "1      VMAPRTLIL  HLA class I histocompatibility antigen, Cw-3 a...   \n",
      "2      FLRGRFYGL                                                NaN   \n",
      "3      FLRGRAYGL                             nuclear antigen EBNA-3   \n",
      "4      EEYLKAWTF                                                NaN   \n",
      "\n",
      "                  Epitope - Source Organism Assay - Type  ...  \\\n",
      "0                      Homo sapiens (human)       T cell  ...   \n",
      "1                      Homo sapiens (human)       T cell  ...   \n",
      "2                                       NaN       T cell  ...   \n",
      "3  Human herpesvirus 4 (Epstein Barr virus)       T cell  ...   \n",
      "4                                       NaN       T cell  ...   \n",
      "\n",
      "  Chain 2 - CDR1 Start Curated Chain 2 - CDR1 End Curated  \\\n",
      "0                          NaN                        NaN   \n",
      "1                          NaN                        NaN   \n",
      "2                          NaN                        NaN   \n",
      "3                          NaN                        NaN   \n",
      "4                          NaN                        NaN   \n",
      "\n",
      "  Chain 2 - CDR1 Start Calculated Chain 2 - CDR1 End Calculated  \\\n",
      "0                            25.0                          29.0   \n",
      "1                            25.0                          29.0   \n",
      "2                            25.0                          29.0   \n",
      "3                            25.0                          29.0   \n",
      "4                            25.0                          29.0   \n",
      "\n",
      "  Chain 2 - CDR2 Curated Chain 2 - CDR2 Calculated  \\\n",
      "0                    NaN                    FVKESK   \n",
      "1                    NaN                    FVKESK   \n",
      "2                    NaN                    FQNEAQ   \n",
      "3                    NaN                    FQNEAQ   \n",
      "4                    NaN                    FQNEAQ   \n",
      "\n",
      "  Chain 2 - CDR2 Start Curated Chain 2 - CDR2 End Curated  \\\n",
      "0                          NaN                        NaN   \n",
      "1                          NaN                        NaN   \n",
      "2                          NaN                        NaN   \n",
      "3                          NaN                        NaN   \n",
      "4                          NaN                        NaN   \n",
      "\n",
      "   Chain 2 - CDR2 Start Calculated Chain 2 - CDR2 End Calculated  \n",
      "0                             47.0                          52.0  \n",
      "1                             47.0                          52.0  \n",
      "2                             47.0                          52.0  \n",
      "3                             47.0                          52.0  \n",
      "4                             47.0                          52.0  \n",
      "\n",
      "[5 rows x 70 columns]\n",
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 211234 entries, 0 to 211233\n",
      "Data columns (total 70 columns):\n",
      " #   Column                           Non-Null Count   Dtype  \n",
      "---  ------                           --------------   -----  \n",
      " 0   Receptor - Group IRI             211234 non-null  object \n",
      " 1   Receptor - IEDB Receptor ID      211234 non-null  int64  \n",
      " 2   Receptor - Reference Name        20333 non-null   object \n",
      " 3   Receptor - Type                  211234 non-null  object \n",
      " 4   Reference - IEDB IRI             211234 non-null  object \n",
      " 5   Epitope - IEDB IRI               211234 non-null  object \n",
      " 6   Epitope - Name                   211234 non-null  object \n",
      " 7   Epitope - Source Molecule        207297 non-null  object \n",
      " 8   Epitope - Source Organism        207297 non-null  object \n",
      " 9   Assay - Type                     211234 non-null  object \n",
      " 10  Assay - IEDB IDs                 211234 non-null  object \n",
      " 11  Assay - MHC Allele Names         211111 non-null  object \n",
      " 12  Chain 1 - Type                   57216 non-null   object \n",
      " 13  Chain 1 - Organism IRI           57216 non-null   object \n",
      " 14  Chain 1 - Nucleotide Sequence    91 non-null      object \n",
      " 15  Chain 1 - Curated V Gene         38939 non-null   object \n",
      " 16  Chain 1 - Calculated V Gene      8615 non-null    object \n",
      " 17  Chain 1 - Curated D Gene         37 non-null      object \n",
      " 18  Chain 1 - Calculated D Gene      0 non-null       float64\n",
      " 19  Chain 1 - Curated J Gene         36000 non-null   object \n",
      " 20  Chain 1 - Calculated J Gene      8455 non-null    object \n",
      " 21  Chain 1 - Protein Sequence       772 non-null     object \n",
      " 22  Chain 1 - Protein IRI            615 non-null     object \n",
      " 23  Chain 1 - CDR3 Curated           56773 non-null   object \n",
      " 24  Chain 1 - CDR3 Calculated        729 non-null     object \n",
      " 25  Chain 1 - CDR3 Start Curated     7 non-null       float64\n",
      " 26  Chain 1 - CDR3 End Curated       7 non-null       float64\n",
      " 27  Chain 1 - CDR3 Start Calculated  729 non-null     float64\n",
      " 28  Chain 1 - CDR3 End Calculated    729 non-null     float64\n",
      " 29  Chain 1 - CDR1 Curated           263 non-null     object \n",
      " 30  Chain 1 - CDR1 Calculated        729 non-null     object \n",
      " 31  Chain 1 - CDR1 Start Curated     7 non-null       float64\n",
      " 32  Chain 1 - CDR1 End Curated       7 non-null       float64\n",
      " 33  Chain 1 - CDR1 Start Calculated  729 non-null     float64\n",
      " 34  Chain 1 - CDR1 End Calculated    729 non-null     float64\n",
      " 35  Chain 1 - CDR2 Curated           270 non-null     object \n",
      " 36  Chain 1 - CDR2 Calculated        729 non-null     object \n",
      " 37  Chain 1 - CDR2 Start Curated     7 non-null       float64\n",
      " 38  Chain 1 - CDR2 End Curated       7 non-null       float64\n",
      " 39  Chain 1 - CDR2 Start Calculated  729 non-null     float64\n",
      " 40  Chain 1 - CDR2 End Calculated    729 non-null     float64\n",
      " 41  Chain 2 - Type                   182103 non-null  object \n",
      " 42  Chain 2 - Organism IRI           182099 non-null  object \n",
      " 43  Chain 2 - Nucleotide Sequence    711 non-null     object \n",
      " 44  Chain 2 - Curated V Gene         159649 non-null  object \n",
      " 45  Chain 2 - Calculated V Gene      12331 non-null   object \n",
      " 46  Chain 2 - Curated D Gene         32021 non-null   object \n",
      " 47  Chain 2 - Calculated D Gene      768 non-null     object \n",
      " 48  Chain 2 - Curated J Gene         157016 non-null  object \n",
      " 49  Chain 2 - Calculated J Gene      12127 non-null   object \n",
      " 50  Chain 2 - Protein Sequence       826 non-null     object \n",
      " 51  Chain 2 - Protein IRI            671 non-null     object \n",
      " 52  Chain 2 - CDR3 Curated           180899 non-null  object \n",
      " 53  Chain 2 - CDR3 Calculated        137692 non-null  object \n",
      " 54  Chain 2 - CDR3 Start Curated     15 non-null      float64\n",
      " 55  Chain 2 - CDR3 End Curated       15 non-null      float64\n",
      " 56  Chain 2 - CDR3 Start Calculated  778 non-null     float64\n",
      " 57  Chain 2 - CDR3 End Calculated    778 non-null     float64\n",
      " 58  Chain 2 - CDR1 Curated           6274 non-null    object \n",
      " 59  Chain 2 - CDR1 Calculated        778 non-null     object \n",
      " 60  Chain 2 - CDR1 Start Curated     7 non-null       float64\n",
      " 61  Chain 2 - CDR1 End Curated       7 non-null       float64\n",
      " 62  Chain 2 - CDR1 Start Calculated  778 non-null     float64\n",
      " 63  Chain 2 - CDR1 End Calculated    778 non-null     float64\n",
      " 64  Chain 2 - CDR2 Curated           6355 non-null    object \n",
      " 65  Chain 2 - CDR2 Calculated        778 non-null     object \n",
      " 66  Chain 2 - CDR2 Start Curated     7 non-null       float64\n",
      " 67  Chain 2 - CDR2 End Curated       7 non-null       float64\n",
      " 68  Chain 2 - CDR2 Start Calculated  778 non-null     float64\n",
      " 69  Chain 2 - CDR2 End Calculated    778 non-null     float64\n",
      "dtypes: float64(25), int64(1), object(44)\n",
      "memory usage: 112.8+ MB\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(df.head())\n",
    "print(df.info())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "interestingProperties = [\n",
    "  \"Receptor - IEDB Receptor ID\",\n",
    "  \"Epitope - Name\",\n",
    "  \"Epitope - Source Organism\",\n",
    "  \"Assay - Type\", \n",
    "  \"Assay - MHC Allele Names\", # maybe interesting because EPIC Trace does something with Allele -> https://github.com/DaniTheOrange/EPIC-TRACE/blob/main/src/construct_long.py\n",
    "  \"Chain 1 - Type\",\n",
    "  \"Chain 1 - Curated V Gene\",\n",
    "  \"Chain 1 - Curated J Gene\",\n",
    "  \"Chain 1 - Protein Sequence\", # this one is a very long chain\n",
    "  \"Chain 1 - CDR3 Curated\", # CDR Sequences also have start and end (curated/calculated properties), which is a integer\n",
    "  \"Chain 1 - CDR1 Curated\",\n",
    "  \"Chain 1 - CDR2 Curated\",\n",
    "  \"Chain 2 - Type\",\n",
    "  \"Chain 2 - Curated V Gene\",\n",
    "  \"Chain 2 - Curated J Gene\",\n",
    "  \"Chain 2 - Protein Sequence\", # this one is a very long chain\n",
    "  \"Chain 2 - CDR3 Curated\",\n",
    "  \"Chain 2 - CDR1 Curated\",\n",
    "  \"Chain 2 - CDR2 Curated\",\n",
    "  ]\n",
    "\n",
    "maybeInterestingProperties = [\n",
    "  \"Chain 1 - Nucleotide Sequence\", \n",
    "  \"Chain 1 - Curated D Gene\", \n",
    "  \"Chain 1 - Calculated D Gene\", \n",
    "  \"Chain 2 - Nucleotide Sequence\", \n",
    "  \"Chain 2 - Curated D Gene\", \n",
    "  \"Chain 2 - Calculated D Gene\"\n",
    "  ]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the leftover attributes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chain 2 - CDR2 Calculated\n",
      "Chain 2 - CDR2 Start Calculated\n",
      "Chain 2 - CDR1 End Curated\n",
      "Chain 1 - CDR3 End Curated\n",
      "Receptor - Group IRI\n",
      "Chain 2 - Calculated V Gene\n",
      "Chain 1 - CDR2 End Calculated\n",
      "Assay - IEDB IDs\n",
      "Chain 1 - CDR1 Start Calculated\n",
      "Chain 2 - CDR3 Start Curated\n",
      "Receptor - Reference Name\n",
      "Chain 2 - CDR2 End Calculated\n",
      "Chain 1 - CDR3 Start Calculated\n",
      "Chain 1 - CDR1 End Calculated\n",
      "Chain 2 - CDR3 End Curated\n",
      "Chain 1 - CDR3 Calculated\n",
      "Chain 2 - CDR2 Start Curated\n",
      "Chain 1 - Calculated V Gene\n",
      "Chain 1 - CDR3 End Calculated\n",
      "Reference - IEDB IRI\n",
      "Chain 1 - CDR2 End Curated\n",
      "Chain 2 - CDR1 End Calculated\n",
      "Chain 1 - CDR1 End Curated\n",
      "Chain 2 - Protein IRI\n",
      "Epitope - IEDB IRI\n",
      "Epitope - Source Molecule\n",
      "Chain 1 - CDR3 Start Curated\n",
      "Chain 2 - Calculated J Gene\n",
      "Chain 1 - Calculated J Gene\n",
      "Chain 1 - Organism IRI\n",
      "Chain 2 - CDR1 Start Curated\n",
      "Chain 2 - CDR3 End Calculated\n",
      "Chain 1 - CDR1 Calculated\n",
      "Chain 2 - Organism IRI\n",
      "Chain 1 - CDR1 Start Curated\n",
      "Chain 2 - CDR1 Start Calculated\n",
      "Chain 2 - CDR2 End Curated\n",
      "Receptor - Type\n",
      "Chain 1 - CDR2 Start Calculated\n",
      "Chain 1 - CDR2 Start Curated\n",
      "Chain 2 - CDR3 Calculated\n",
      "Chain 2 - CDR3 Start Calculated\n",
      "Chain 1 - CDR2 Calculated\n",
      "Chain 2 - CDR1 Calculated\n",
      "Chain 1 - Protein IRI\n"
     ]
    }
   ],
   "source": [
    "allProperties = df.columns.to_list()\n",
    "\n",
    "# Convert lists to sets for easier computation\n",
    "all_properties_set = set(allProperties)\n",
    "interesting_properties_set = set(interestingProperties)\n",
    "maybe_interesting_properties_set = set(maybeInterestingProperties)\n",
    "\n",
    "# Combine interesting and maybe interesting properties\n",
    "combined_interesting_set = interesting_properties_set.union(maybe_interesting_properties_set)\n",
    "\n",
    "# Find the difference\n",
    "leftoverProperties = list(all_properties_set - combined_interesting_set)\n",
    "\n",
    "# Print or return the leftover properties\n",
    "for property in leftoverProperties:\n",
    "    print(property)\n",
    "#to be checked (also if CDR Start/End is needed, look at table columns to know what it's meaning):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Export the most interesting data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('../../data/IEDB/IEDB.csv', index=False, columns=interestingProperties)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "BA24",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
